src/instana/instrumentation/kafka/confluent_kafka_python.py

```diff
@@ -105,78 +113,142 @@ def create_span(
+                    if tracer.exporter.options.kafka_trace_correlation
+                    else None
+                )
+            )
+            span = tracer.start_span(
+                "kafka-consumer", span_context=parent_context, kind=SpanKind.CONSUMER
             )
-        )
-        with tracer.start_as_current_span(
-            "kafka-consumer", span_context=parent_context, kind=SpanKind.CONSUMER
-        ) as span:
             if topic:
                 span.set_attribute("kafka.service", topic)
             span.set_attribute("kafka.access", span_type)
-
             if exception:
                 span.record_exception(exception)
+                span.end()
+
+            save_consumer_span_into_context(span)
+        except Exception as e:
+            logger.debug(f"Error while creating kafka-consumer span: {e}")
+
+    def save_consumer_span_into_context(span: "InstanaSpan") -> None:
+        ctx = trace.set_span_in_context(span)
+        token = context.attach(ctx)
+        consumer_token["token"] = token
+        consumer_span.set(span)
+
+    def close_consumer_span(span: "InstanaSpan") -> None:
+        if span.is_recording():
+            span.end()
+            consumer_span.set(None)
+        if "token" in consumer_token:
+            context.detach(consumer_token.pop("token", None))
+
+    def clear_context() -> None:
+        context.attach(trace.set_span_in_context(None))
+        consumer_token.clear()
+        consumer_span.set(None)
 
     def trace_kafka_consume(
         wrapped: Callable[..., InstanaConfluentKafkaConsumer.consume],
         instance: InstanaConfluentKafkaConsumer,
         args: Tuple[int, str, Tuple[Any, ...]],
         kwargs: Dict[str, Any],
     ) -> List[confluent_kafka.Message]:
-        if tracing_is_off():
-            return wrapped(*args, **kwargs)
-
         res = None
         exception = None
 
         try:
             res = wrapped(*args, **kwargs)
+            for message in res:
+                create_span("consume", message.topic(), message.headers())
+            return res
         except Exception as exc:
             exception = exc
-        finally:
-            if res:
-                for message in res:
-                    create_span("consume", message.topic(), message.headers())
-            else:
-                create_span("consume", exception=exception)
+            create_span("consume", exception=exception)
+
+    def trace_kafka_close(
+        wrapped: Callable[..., InstanaConfluentKafkaConsumer.close],
+        instance: InstanaConfluentKafkaConsumer,
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> None:
+        try:
+            # Close any existing consumer span before closing the consumer
+            span = consumer_span.get(None)
+            if span is not None:
+                close_consumer_span(span)
+
+            # Execute the actual close operation
+            result = wrapped(*args, **kwargs)
```

Comment:
Use the variable name `res` to keep compatibility with the rest of the code.
```suggestion
            res = wrapped(*args, **kwargs)
```